Failure looks related:

```
2020-08-21T01:07:48.2222974Z [INFO] Running io.quarkus.it.spring.web.ExceptionHandlingTest
2020-08-21T01:07:53.2554499Z [ERROR] Tests run: 18, Failures: 0, Errors: 1, Skipped: 17, Time elapsed: 5.016 s <<< FAILURE! - in io.quarkus.it.spring.web.ExceptionHandlingTest
2020-08-21T01:07:53.2555409Z [ERROR] io.quarkus.it.spring.web.ExceptionHandlingTest.testUnhandledAnnotatedException  Time elapsed: 0.013 s  <<< ERROR!
2020-08-21T01:07:53.2556172Z java.lang.RuntimeException: 
2020-08-21T01:07:53.2561661Z java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2020-08-21T01:07:53.2563752Z 	[error]: Build step io.quarkus.hibernate.validator.deployment.HibernateValidatorProcessor#build threw an exception: java.lang.IllegalArgumentException: Not a method
2020-08-21T01:07:53.2564427Z 	at org.jboss.jandex.ClassInfo.asMethod(ClassInfo.java:611)
2020-08-21T01:07:53.2566323Z 	at io.quarkus.hibernate.validator.deployment.HibernateValidatorProcessor.gatherJaxRsMethods(HibernateValidatorProcessor.java:362)
2020-08-21T01:07:53.2568112Z 	at io.quarkus.hibernate.validator.deployment.HibernateValidatorProcessor.build(HibernateValidatorProcessor.java:251)
2020-08-21T01:07:53.2568861Z 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-08-21T01:07:53.2570134Z 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-08-21T01:07:53.2571713Z 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-08-21T01:07:53.2572370Z 	at java.lang.reflect.Method.invoke(Method.java:498)
2020-08-21T01:07:53.2573493Z 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:936)
2020-08-21T01:07:53.2574286Z 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
2020-08-21T01:07:53.2575744Z 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
2020-08-21T01:07:53.2576915Z 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
2020-08-21T01:07:53.2578307Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
2020-08-21T01:07:53.2579587Z 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
2020-08-21T01:07:53.2580052Z 	at java.lang.Thread.run(Thread.java:748)
2020-08-21T01:07:53.2581294Z 	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
2020-08-21T01:07:53.2581607Z 
```

--- a/.github/workflows/ci-actions.yml
+++ b/.github/workflows/ci-actions.yml
@@ -117,12 +117,11 @@ jobs:
             java-version: 11,
             maven_args: "-pl !integration-tests/gradle"
           }
-          # Disabled for now as totally unstable
-          #- {
-          #  name: Java 14,
-          #  java-version: 14,
-          #  maven_args: "-pl !integration-tests/gradle"
-          #}
+          - {
+            name: Java 14,
+            java-version: 14,
+            maven_args: "-pl !integration-tests/gradle"
+          }
 
     services:
       keycloak:

--- a/extensions/hibernate-validator/deployment/pom.xml
+++ b/extensions/hibernate-validator/deployment/pom.xml
@@ -24,6 +24,10 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-common-spi</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-resteasy-server-common-spi</artifactId>
         </dependency>
         <dependency>

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
@@ -5,7 +5,6 @@ import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 import java.lang.annotation.Repeatable;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -66,6 +65,7 @@ import io.quarkus.hibernate.validator.runtime.HibernateValidatorBuildTimeConfig;
 import io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder;
 import io.quarkus.hibernate.validator.runtime.ValidatorProvider;
 import io.quarkus.hibernate.validator.runtime.interceptor.MethodValidationInterceptor;
+import io.quarkus.resteasy.common.spi.ResteasyDotNames;
 import io.quarkus.resteasy.server.common.spi.AdditionalJaxRsResourceMethodAnnotationsBuildItem;
 import io.quarkus.runtime.LocalesBuildTimeConfig;
 
@@ -93,16 +93,6 @@ class HibernateValidatorProcessor {
     private static final DotName VALID = DotName.createSimple(Valid.class.getName());
 
     private static final DotName REPEATABLE = DotName.createSimple(Repeatable.class.getName());
-
-    private static final DotName[] JAXRS_METHOD_ANNOTATIONS = {
-            DotName.createSimple("javax.ws.rs.GET"),
-            DotName.createSimple("javax.ws.rs.HEAD"),
-            DotName.createSimple("javax.ws.rs.DELETE"),
-            DotName.createSimple("javax.ws.rs.OPTIONS"),
-            DotName.createSimple("javax.ws.rs.PATCH"),
-            DotName.createSimple("javax.ws.rs.POST"),
-            DotName.createSimple("javax.ws.rs.PUT"),
-    };
 
     private static final Pattern BUILT_IN_CONSTRAINT_REPEATABLE_CONTAINER_PATTERN = Pattern.compile("\\$List$");
 
@@ -349,8 +339,8 @@ class HibernateValidatorProcessor {
         Map<DotName, Set<SimpleMethodSignatureKey>> jaxRsMethods = new HashMap<>();
 
         Collection<DotName> jaxRsMethodDefiningAnnotations = new ArrayList<>(
-                JAXRS_METHOD_ANNOTATIONS.length + additionalJaxRsResourceMethodAnnotations.size());
-        jaxRsMethodDefiningAnnotations.addAll(Arrays.asList(JAXRS_METHOD_ANNOTATIONS));
+                ResteasyDotNames.JAXRS_METHOD_ANNOTATIONS.size() + additionalJaxRsResourceMethodAnnotations.size());
+        jaxRsMethodDefiningAnnotations.addAll(ResteasyDotNames.JAXRS_METHOD_ANNOTATIONS);
         for (AdditionalJaxRsResourceMethodAnnotationsBuildItem additionalJaxRsResourceMethodAnnotation : additionalJaxRsResourceMethodAnnotations) {
             jaxRsMethodDefiningAnnotations.addAll(additionalJaxRsResourceMethodAnnotation.getAnnotationClasses());
         }

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
@@ -4,6 +4,8 @@ import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 
 import java.lang.annotation.Repeatable;
 import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
@@ -91,6 +93,16 @@ class HibernateValidatorProcessor {
     private static final DotName VALID = DotName.createSimple(Valid.class.getName());
 
     private static final DotName REPEATABLE = DotName.createSimple(Repeatable.class.getName());
+
+    private static final DotName[] JAXRS_METHOD_ANNOTATIONS = {
+            DotName.createSimple("javax.ws.rs.GET"),
+            DotName.createSimple("javax.ws.rs.HEAD"),
+            DotName.createSimple("javax.ws.rs.DELETE"),
+            DotName.createSimple("javax.ws.rs.OPTIONS"),
+            DotName.createSimple("javax.ws.rs.PATCH"),
+            DotName.createSimple("javax.ws.rs.POST"),
+            DotName.createSimple("javax.ws.rs.PUT"),
+    };
 
     private static final Pattern BUILT_IN_CONSTRAINT_REPEATABLE_CONTAINER_PATTERN = Pattern.compile("\\$List$");
 
@@ -188,7 +200,7 @@ class HibernateValidatorProcessor {
         consideredAnnotations.add(VALIDATE_ON_EXECUTION);
 
         Set<DotName> classNamesToBeValidated = new HashSet<>();
-        Map<DotName, Set<String>> inheritedAnnotationsToBeValidated = new HashMap<>();
+        Map<DotName, Set<SimpleMethodSignatureKey>> inheritedAnnotationsToBeValidated = new HashMap<>();
         Set<String> detectedBuiltinConstraints = new HashSet<>();
 
         for (DotName consideredAnnotation : consideredAnnotations) {
@@ -236,15 +248,18 @@ class HibernateValidatorProcessor {
             }
         }
 
+        // JAX-RS methods are handled differently by the transformer so those need to be gathered here.
+        // Note: The focus only on methods is basically an incomplete solution, since there could also be
+        // class-level JAX-RS annotations but currently the transformer only looks at methods.
+        Map<DotName, Set<SimpleMethodSignatureKey>> jaxRsMethods = gatherJaxRsMethods(additionalJaxRsResourceMethodAnnotations,
+                indexView);
+
         // Add the annotations transformer to add @MethodValidated annotations on the methods requiring validation
-        Set<DotName> additionalJaxRsMethodAnnotationsDotNames = new HashSet<>(additionalJaxRsResourceMethodAnnotations.size());
-        for (AdditionalJaxRsResourceMethodAnnotationsBuildItem additionalJaxRsResourceMethodAnnotation : additionalJaxRsResourceMethodAnnotations) {
-            additionalJaxRsMethodAnnotationsDotNames.addAll(additionalJaxRsResourceMethodAnnotation.getAnnotationClasses());
-        }
+
         annotationsTransformers
                 .produce(new AnnotationsTransformerBuildItem(
                         new MethodValidatedAnnotationsTransformer(consideredAnnotations,
-                                additionalJaxRsMethodAnnotationsDotNames,
+                                jaxRsMethods,
                                 inheritedAnnotationsToBeValidated)));
 
         Set<Class<?>> classesToBeValidated = new HashSet<>();
@@ -317,14 +332,45 @@ class HibernateValidatorProcessor {
         }
     }
 
-    private static void contributeMethodsWithInheritedValidation(Map<DotName, Set<String>> inheritedAnnotationsToBeValidated,
+    private static void contributeMethodsWithInheritedValidation(
+            Map<DotName, Set<SimpleMethodSignatureKey>> inheritedAnnotationsToBeValidated,
             IndexView indexView, MethodInfo method) {
         ClassInfo clazz = method.declaringClass();
         if (Modifier.isInterface(clazz.flags())) {
             // Remember annotated interface methods that must be validated
-            inheritedAnnotationsToBeValidated.computeIfAbsent(clazz.name(), k -> new HashSet<String>())
-                    .add(method.name().toString());
+            inheritedAnnotationsToBeValidated.computeIfAbsent(clazz.name(), k -> new HashSet<>())
+                    .add(new SimpleMethodSignatureKey(method));
         }
+    }
+
+    private static Map<DotName, Set<SimpleMethodSignatureKey>> gatherJaxRsMethods(
+            List<AdditionalJaxRsResourceMethodAnnotationsBuildItem> additionalJaxRsResourceMethodAnnotations,
+            IndexView indexView) {
+        Map<DotName, Set<SimpleMethodSignatureKey>> jaxRsMethods = new HashMap<>();
+
+        Collection<DotName> jaxRsMethodDefiningAnnotations = new ArrayList<>(
+                JAXRS_METHOD_ANNOTATIONS.length + additionalJaxRsResourceMethodAnnotations.size());
+        jaxRsMethodDefiningAnnotations.addAll(Arrays.asList(JAXRS_METHOD_ANNOTATIONS));
+        for (AdditionalJaxRsResourceMethodAnnotationsBuildItem additionalJaxRsResourceMethodAnnotation : additionalJaxRsResourceMethodAnnotations) {
+            jaxRsMethodDefiningAnnotations.addAll(additionalJaxRsResourceMethodAnnotation.getAnnotationClasses());
+        }
+
+        for (DotName jaxRsAnnotation : jaxRsMethodDefiningAnnotations) {
+            Collection<AnnotationInstance> annotationInstances = indexView.getAnnotations(jaxRsAnnotation);
+
+            if (annotationInstances.isEmpty()) {
+                continue;
+            }
+
+            for (AnnotationInstance annotation : annotationInstances) {
+                if (annotation.target().kind() == AnnotationTarget.Kind.METHOD) {
+                    MethodInfo method = annotation.target().asMethod();
+                    jaxRsMethods.computeIfAbsent(method.declaringClass().name(), k -> new HashSet<>())
+                            .add(new SimpleMethodSignatureKey(method));
+                }
+            }
+        }
+        return jaxRsMethods;
     }
 
     private static DotName getClassName(Type type) {
@@ -342,4 +388,5 @@ class HibernateValidatorProcessor {
     private static boolean hasXmlConfiguration() {
         return Thread.currentThread().getContextClassLoader().getResource(META_INF_VALIDATION_XML) != null;
     }
+
 }

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/MethodValidatedAnnotationsTransformer.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/MethodValidatedAnnotationsTransformer.java
@@ -1,9 +1,6 @@
 package io.quarkus.hibernate.validator.deployment;
 
 import java.lang.reflect.Modifier;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
 
@@ -14,6 +11,7 @@ import org.jboss.jandex.MethodInfo;
 import org.jboss.logging.Logger;
 
 import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.deployment.index.IndexingUtil;
 import io.quarkus.hibernate.validator.runtime.interceptor.MethodValidated;
 import io.quarkus.hibernate.validator.runtime.jaxrs.JaxrsEndPointValidated;
 
@@ -24,30 +22,16 @@ public class MethodValidatedAnnotationsTransformer implements AnnotationsTransfo
 
     private static final Logger LOGGER = Logger.getLogger(MethodValidatedAnnotationsTransformer.class.getPackage().getName());
 
-    private static final DotName[] JAXRS_METHOD_ANNOTATIONS = {
-            DotName.createSimple("javax.ws.rs.GET"),
-            DotName.createSimple("javax.ws.rs.HEAD"),
-            DotName.createSimple("javax.ws.rs.DELETE"),
-            DotName.createSimple("javax.ws.rs.OPTIONS"),
-            DotName.createSimple("javax.ws.rs.PATCH"),
-            DotName.createSimple("javax.ws.rs.POST"),
-            DotName.createSimple("javax.ws.rs.PUT"),
-    };
-
     private final Set<DotName> consideredAnnotations;
-    private final Collection<DotName> effectiveJaxRsMethodDefiningAnnotations;
-    private final Map<DotName, Set<String>> inheritedAnnotationsToBeValidated;
+    private final Map<DotName, Set<SimpleMethodSignatureKey>> inheritedAnnotationsToBeValidated;
+    private final Map<DotName, Set<SimpleMethodSignatureKey>> jaxRsMethods;
 
     MethodValidatedAnnotationsTransformer(Set<DotName> consideredAnnotations,
-            Collection<DotName> additionalJaxRsMethodAnnotationsDotNames,
-            Map<DotName, Set<String>> inheritedAnnotationsToBeValidated) {
+            Map<DotName, Set<SimpleMethodSignatureKey>> jaxRsMethods,
+            Map<DotName, Set<SimpleMethodSignatureKey>> inheritedAnnotationsToBeValidated) {
         this.consideredAnnotations = consideredAnnotations;
+        this.jaxRsMethods = jaxRsMethods;
         this.inheritedAnnotationsToBeValidated = inheritedAnnotationsToBeValidated;
-
-        this.effectiveJaxRsMethodDefiningAnnotations = new ArrayList<>(
-                JAXRS_METHOD_ANNOTATIONS.length + additionalJaxRsMethodAnnotationsDotNames.size());
-        effectiveJaxRsMethodDefiningAnnotations.addAll(Arrays.asList(JAXRS_METHOD_ANNOTATIONS));
-        effectiveJaxRsMethodDefiningAnnotations.addAll(additionalJaxRsMethodAnnotationsDotNames);
     }
 
     @Override
@@ -85,10 +69,12 @@ public class MethodValidatedAnnotationsTransformer implements AnnotationsTransfo
         }
         // This method has no annotations of its own: look for inherited annotations
         ClassInfo clazz = method.declaringClass();
-        String methodName = method.name().toString();
-        for (Map.Entry<DotName, Set<String>> validatedMethod : inheritedAnnotationsToBeValidated.entrySet()) {
-            if (clazz.interfaceNames().contains(validatedMethod.getKey())
-                    && validatedMethod.getValue().contains(methodName)) {
+        SimpleMethodSignatureKey signatureKey = new SimpleMethodSignatureKey(method);
+        for (Map.Entry<DotName, Set<SimpleMethodSignatureKey>> validatedMethod : inheritedAnnotationsToBeValidated.entrySet()) {
+            DotName ifaceOrSuperClass = validatedMethod.getKey();
+            // note: only check the direct superclass since we do not (yet) have the entire ClassInfo hierarchy here
+            if ((clazz.interfaceNames().contains(ifaceOrSuperClass) || ifaceOrSuperClass.equals(clazz.superName()))
+                    && validatedMethod.getValue().contains(signatureKey)) {
                 return true;
             }
         }
@@ -96,11 +82,32 @@ public class MethodValidatedAnnotationsTransformer implements AnnotationsTransfo
     }
 
     private boolean isJaxrsMethod(MethodInfo method) {
-        for (DotName jaxrsMethodAnnotation : effectiveJaxRsMethodDefiningAnnotations) {
-            if (method.hasAnnotation(jaxrsMethodAnnotation)) {
+        ClassInfo clazz = method.declaringClass();
+        SimpleMethodSignatureKey signatureKey = new SimpleMethodSignatureKey(method);
+
+        if (isJaxrsMethod(signatureKey, clazz.name())) {
+            return true;
+        }
+
+        // check interfaces
+        for (DotName iface : clazz.interfaceNames()) {
+            if (isJaxrsMethod(signatureKey, iface)) {
+                return true;
+            }
+        }
+
+        // check superclass, but only the direct one since we do not (yet) have the entire ClassInfo hierarchy here
+        DotName superClass = clazz.superName();
+        if (!superClass.equals(IndexingUtil.OBJECT)) {
+            if (isJaxrsMethod(signatureKey, superClass)) {
                 return true;
             }
         }
         return false;
+    }
+
+    private boolean isJaxrsMethod(SimpleMethodSignatureKey signatureKey, DotName dotName) {
+        Set<SimpleMethodSignatureKey> signatureKeys = jaxRsMethods.get(dotName);
+        return signatureKeys != null && signatureKeys.contains(signatureKey);
     }
 }

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/SimpleMethodSignatureKey.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/SimpleMethodSignatureKey.java
@@ -1,0 +1,44 @@
+package io.quarkus.hibernate.validator.deployment;
+
+import java.util.stream.Collectors;
+
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.Type;
+
+class SimpleMethodSignatureKey {
+
+    private final String key;
+
+    SimpleMethodSignatureKey(MethodInfo method) {
+        // Notes:
+        // - MethodInfo.toString() is not usable here because it includes the declaring class
+        // - just parameters() for the second part would include annotations (see Type.toString())
+        key = method.name() + method.parameters().stream()
+                .map(Type::name)
+                .map(DotName::toString)
+                .collect(Collectors.joining(", ", "(", ")"));
+    }
+
+    @Override
+    public int hashCode() {
+        return key.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        SimpleMethodSignatureKey other = (SimpleMethodSignatureKey) obj;
+        return key.equals(other.key);
+    }
+
+    @Override
+    public String toString() {
+        return key;
+    }
+}

--- a/integration-tests/hibernate-validator/src/test/java/io/quarkus/it/hibernate/validator/HibernateValidatorFunctionalityTest.java
+++ b/integration-tests/hibernate-validator/src/test/java/io/quarkus/it/hibernate/validator/HibernateValidatorFunctionalityTest.java
@@ -76,8 +76,7 @@ public class HibernateValidatorFunctionalityTest {
         RestAssured.when()
                 .get("/hibernate-validator/test/rest-end-point-interface-validation/plop/")
                 .then()
-                // 500 until this is fixed: https://github.com/quarkusio/quarkus/issues/11341#issuecomment-673146350
-                .statusCode(500)
+                .statusCode(400)
                 .body(containsString("numeric value out of bounds"));
 
         RestAssured.when()
@@ -91,8 +90,7 @@ public class HibernateValidatorFunctionalityTest {
         RestAssured.when()
                 .get("/hibernate-validator/test/rest-end-point-interface-validation-annotation-on-impl-method/plop/")
                 .then()
-                // 500 until this is fixed: https://github.com/quarkusio/quarkus/issues/11341#issuecomment-673146350
-                .statusCode(500)
+                .statusCode(400)
                 .body(containsString("numeric value out of bounds"));
 
         RestAssured.when()

